Fixed #25774 -- Added Extract and Trunc datetime functions.

jarshwah opened the pull request:
A bunch of datetime transforms/expression are moved and renamed into their own module underneath `django.db.models.functions.datetime`. I'm open to relocating them elsewhere, but I thought it'd be nice to group them within their own file.

I do plan to implement Trunc expressions too. There's already a high level Date and DateTime type within `expressions.py` which is only used for `.datetimes()` and `.dates()` queryset methods. I might retire those in favour of new public expressions if it's easy enough to do.

Todo:
- [x] Move and rename existing date/time transforms into public module with public names
- [x] Docs
- [x] Tests
- [x] Implement TruncX in style of ExtractX (with docs/tests) 
- [x] Release notes
- [x] Fix test failures if pytz isn't installed
- [x] Check tests work on Windows without pytz
- [x] Check Oracle failures (test_trunc_func, test_trunc_second_func)

timgraham commented:
Listing all functions in one document (database-functions.txt) seems okay to me.

charettes commented on django/db/models/lookups.py:
```diff
@@ -489,7 +433,19 @@ def get_bound(self):
+        params.extend(bounds)
+        return '%s BETWEEN %%s AND %%s' % lhs_sql, params
+
+
```
We should make sure the `YearLookup` subclasses are registered to the `ExtractYear` transform as they perform operations that can use indexes.

charettes replied:
We should test that `datefield__year__(exact|gt|gte|lt|lte)` actually uses these lookups as the suite pass with them unregistered.

jarshwah replied:
They are registered against YearLookup (https://github.com/django/django/pull/6243/files#diff-d4a6ef1ea81ea8f2b3d3c53d39463b90R92) but I'll make sure there are tests that verify the behaviour. Thanks for pointing that out.

jarshwah commented:
Bikeshedding: is django/db/models/functions/date.py ok? I originally had it named `datetime.py` but that broke imports of the builtin datetime module which led me to change it. I could also name the file `datetimes.py` but I'm not sure that's better. Anyone have a better name?

timgraham commented:
If you want to use `datetime.py` you can probably just add `from __future__ import absolute_import` to fix the import issues.

charettes commented on django/db/models/functions/date.py:
```diff
@@ -0,0 +1,216 @@
+
+
+class TruncDay(TruncDateField):
+    lookup_name = 'date'
```
Should this be `kind = 'day'`?

charettes replied:
Oh I see it's the `datetimefield__date` lookup. Do you think `TruncDate` would make more sense for a class name here?

charettes replied:
NVM since it's both used as the `__date` lookup and as but truncation API the name sounds good.

charettes commented:
This is looking really good. I think the tests should be run with and without timezone enabled which is not the case actually as the CI settings set `USE_TZ=True`. I'd suggest you declare a subclass of your actual test case class with `override_settings(USE_TZ=True, TIMEZONE='...')` and the former with `override_settings(USE_TZ=False)`.

jarshwah commented:
Yeah very good idea. I was wondering how I was going to test timezone support here. The trunc classes also take a tzinfo attribute which I haven't tested either. I've got a few more tests to write, so I probably shouldn't have ticked the test todo box at the top.

jarshwah commented:
I'm having a bit of trouble handling user defined tzinfo, and what the actual return types should be. 

Given this query:

```
tzinfo = pytz.timezone('Australia/Melbourne') # +10:00 during June, +11:00 during January
# start_date = '2015-06-01 14:00:00' UTC
qs = DTModel.objects.annotate(
     truncated=Trunc('start_date', kind, output_field=DateField(), tzinfo=tzinfo)
).order_by('start_datetime')
model = qs.first()
print(model.truncated) # ??
```
What should `model.truncated` return?

1) '2015-01-01 00:00:00' in UTC?
2) '2015-01-01 00:00:00' in +10:00 (timezone of 2015-06-01)
3) '2015-01-01 00:00:00' in +11:00 (timezone of 2015-01-01)

The reason this matters is because if a user passes in a datetime with +10:00 and gets a return value with +11:00, the user is going to need to either create a brand new datetime from scratch using the appropriate timezone, or a combination of make_aware(make_naive(value)) in various offsets if they want to do comparisons.

The existing converter is also probably wrong for truncated types:

```
value = value.replace(tzinfo=None)
value = timezone.make_aware(value, self.tzinfo) # now the truncated to year time has an hour of 11!
```

There is now a time component for a trunc to year, but all values will have the same time component, AND the user asked for time in a specific timezone. I'm not sure how to resolve this.

jarshwah commented:
I went with option 3) which the converter had already been doing. The returned value will have the offset of the datetime after truncation. That is if we truncate a time in offset +10:00 which results in a datetime which would be +11:00, we return offset +11:00.

Dates will always have an appropriate offset, but users will need to be aware that the timezone they pass in can change based on the result of the calculation. I'll call this out specifically in the docs.

jarshwah commented:
I've removed the `tzinfo` parameter from the `Trunc` class. I can't think of a reason why users should need to use that parameter, so I simplified it. The internals do a `timezone.get_current_timezone()` anyway which should always do the right thing. Can always re-add later if someone provides a use case that requires it. For future sake, below is a copy of the test that I removed incase it needs to be added later.

The docs are finished but I still need to write release notes. Most of the docs were copy/paste/edit, but I've tested all the example code in a repl session and it seems fine.

I'm going to wait for #6147 before merging this one because that's almost ready to go, and I don't want that author to have to rebase yet again.

Removed test code:

```
@skipIf(pytz is None, "this test requires pytz")
    def test_trunc_func_with_timezone(self):
        """
        If the truncated datetime transitions to a different offset (day light
        savings) then the returned value will have that new timezone/offset.
        """
        start_datetime = microsecond_support(datetime(2015, 6, 15, 14, 30, 50, 321))
        end_datetime = microsecond_support(datetime(2016, 6, 15, 14, 10, 50, 123))
        start_datetime = timezone.make_aware(start_datetime, is_dst=False)
        end_datetime = timezone.make_aware(end_datetime, is_dst=False)
        self.create_model(start_datetime, end_datetime)
        self.create_model(end_datetime, start_datetime)

        tzinfo = pytz.timezone('Australia/Melbourne')

        def test_datetime_kind(kind, tzinfo=tzinfo):
            self.assertQuerysetEqual(
                DTModel.objects.annotate(
                    truncated=Trunc('start_datetime', kind, output_field=DateTimeField(), tzinfo=tzinfo)
                ).order_by('start_datetime'),
                [
                    (start_datetime, truncate_to(start_datetime.astimezone(tzinfo), kind, tzinfo)),
                    (end_datetime, truncate_to(end_datetime.astimezone(tzinfo), kind, tzinfo))
                ],
                lambda m: (m.start_datetime, m.truncated)
            )

        def test_date_kind(kind, tzinfo=tzinfo):
            self.assertQuerysetEqual(
                DTModel.objects.annotate(
                    truncated=Trunc('start_date', kind, output_field=DateField(), tzinfo=tzinfo)
                ).order_by('start_datetime'),
                [
                    (start_datetime, truncate_to(start_datetime.date(), kind)),
                    (end_datetime, truncate_to(end_datetime.date(), kind))
                ],
                lambda m: (m.start_datetime, m.truncated)
            )

        test_date_kind('year')
        test_date_kind('month')
        test_date_kind('day')
        test_datetime_kind('year')
        test_datetime_kind('month')
        test_datetime_kind('day')
        test_datetime_kind('hour')
        test_datetime_kind('minute')
        test_datetime_kind('second')

        self.assertEqual(DTModel.objects.filter(
            start_datetime__date=Trunc('start_datetime', 'day', output_field=DateField())).count(), 2)
```

charettes commented on django/db/models/query.py:
```diff
@@ -767,7 +768,7 @@ def datetimes(self, field_name, kind, order='ASC', tzinfo=None):
             tzinfo = None
         return self.annotate(
-            datetimefield=DateTime(field_name, kind, tzinfo),
+            datetimefield=Trunc(field_name, kind, output_field=DateTimeField(), tzinfo=tzinfo),
```
I guess the whole `tzinfo` logic can be removed from here?

jarshwah replied:
Hmm nope, forgot to check that. Guess I have to put the logic back in, but it'll be added as a private implementation detail to support `queryset.datetimes()`.

timgraham commented on django/db/models/functions/__init__.py:
```diff
@@ -4,6 +4,7 @@
 from django.db.models import (
     DateTimeField, Func, IntegerField, Transform, Value,
 )
+from django.db.models.functions.date import *  # NOQA
```
I'd rather not use `import *` since it prevents detecting unused imports. Alternatively, you could move functions.py -> base.py or something and make ` __init__.py` only contains imports.

jarshwah replied:
I've changed it but I'm not sure I've done the right thing. I'm importing all necessary types wrapped up in a `# NOQA`. Would you prefer I use your suggestion of moving all imports to init and current functions to base?

timgraham commented on django/db/models/functions/date.py:
```diff
@@ -0,0 +1,215 @@
+        if self.lookup_name is None:
+            self.lookup_name = lookup_name
+        if self.lookup_name is None:
+            # if we still don't have a lookup name that's an error
```
obvious comment probably not needed

timgraham commented on django/db/models/functions/date.py:
```diff
@@ -0,0 +1,215 @@
+
+class ExtractWeekDay(Extract):
+    """
+    returns Sunday=1 through to Saturday=7.
```
Return ... (pep257 verb style)
chop "to"

timgraham commented on django/db/models/functions/date.py:
```diff
@@ -0,0 +1,215 @@
+    """
+    returns Sunday=1 through to Saturday=7.
+
+    To replicate this in python:
```
Python

timgraham commented on django/db/models/functions/date.py:
```diff
@@ -0,0 +1,215 @@
+
+    def as_sql(self, compiler, connection):
+        inner_sql, inner_params = compiler.compile(self.lhs)
+        # we need to escape any params because trunc_sql will format our string
```
`# Escape any....`

timgraham commented on django/db/models/lookups.py:
```diff
@@ -489,7 +431,28 @@ def get_bound(self):
+        lhs_sql, params = self.process_lhs(compiler, connection, self.lhs.lhs)
+        rhs_sql, rhs_params = self.process_rhs(compiler, connection)
+        try:
+            # checks that rhs_params[0] exists (IndexError),
```
`# Check that ...`

timgraham commented on django/db/models/lookups.py:
```diff
@@ -489,7 +431,28 @@ def get_bound(self):
+            int(rhs_params[0])
+        except (IndexError, TypeError, ValueError):
+            # We can't determine the bounds before executing the query, so
+            # we skip optimisations by falling back to a standard exact
```
optimizations

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -275,3 +275,518 @@ Usage example::
+
+    >>> from datetime import datetime
+    >>> from django.db.models.functions import Extract
+    >>> start = datetime(2015, 6, 15, 14, 30, 50, 321)
```
To simplify the examples, any objection to removing hours/minutes/seconds/microseconds when not needed?

jarshwah replied:
That's fine, I changed all of the Extract examples. I left the Trunc examples alone though, because I think it's useful to show all fields being zeroed out. I can change if you think it'd be better to remove time components from date based truncs.

timgraham commented on django/db/models/functions/__init__.py:
```diff
@@ -1,7 +1,13 @@
 """
```
I think ideally this file would be `base.py` and `__init__.py` would be modeled after https://github.com/django/django/blob/353d436e7cb33cb832a3e8c74b051e3d2ba76018/django/core/files/__init__.py

timgraham commented on django/db/models/functions/date.py:
```diff
@@ -0,0 +1,214 @@
+from datetime import datetime
```
Renaming this file to `datetime.py` and adding `from __future__ import absolute_import` at the top to prevent collisions with the built-in module on Python 2 is working for me. Did you see/consider that suggestion? I do think that name is more intuitive.

aaugustin replied:
So do I.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+``Extract``
+-----------
+
+.. class:: Extract(date_expression, lookup_name=None, **extra)
```
The argument is named `expression` in the code. Did you intentionally rename it in the docs or is it an oversight?

jarshwah replied:
I intentionally renamed it to make the docs clearer that the type of expression should resolve to a date. However, it'd be more accurate to rename to `datetime_expression` so that it fits with functions that only accept a datetime, like Hour/Minute/Second. 

Alternatively, I can rename to `expression` since the following paragraph clarifies the type that expression needs to resolve to. I think that might actually be nicer to read.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+    >>> from django.db.models.functions import Extract
+    >>> start = datetime(2015, 6, 15)
+    >>> end = datetime(2015, 7, 2)
+    >>> Experiment.objects.create(start_time=start, end_time=end)
```
This is giving: `RuntimeWarning: DateTimeField Experiment.end_time received a naive datetime (2015-07-02 00:00:00) while time zone support is active.` Would avoiding that make the examples unnecessarily complex?

aaugustin replied:
Only when `USE_TZ` = True -- but Django encourages that by putting it in the default project template.

jarshwah replied:
It'll add extra noise to add the timezone to each datetime created, plus an extra import. Encouraging good timezone practises in our own docs is probably a good thing though, so I'll make the change.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+
+Usage example::
+
+    >>> from datetime import datetime
```
Given all these Extract functions work similarly, I'm wondering if there's much value in having what are essentially copy/paste examples for every one. It's quite tedious to review.

aaugustin replied:
Yes, I think the documentation would be more clear if you factored:

- extracts that operate on dates and datetimes (year / month / day / week_day)
- extracts that operate only on datetimes (hour / minute / second)

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+Takes a single ``date_expression`` representing a ``DateField`` or
+``DateTimeField`` and a ``lookup_name``, and returns the part of the date
+referenced by ``lookup_name`` as an ``IntegerField``. Django will generally use
+the databases Extract function, so ``lookup_name``\s that aren't explicitly
```
missing apostroph after "databases"

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+* "week_day": From 1 (Sunday) to 7 (Saturday)
+* "hour": 14
+* "minute": 30
+* "second": 50
```
You're assuming that `USE_TZ = False` here. You should clarify that and also give an example with `USE_TZ = True`.

jarshwah replied:
@aaugustin Other than representing the datetime without a timezone component, would USE_TZ affect the actual results for Extract? I understand how timezones affect Trunc and need to add an example for that, was this a misplaced comment?

aaugustin replied:
Yes, it will, because a given point in time may have a different year, month, day, week_day, hour or minute depending on the timezone it's represented in. Differences in hour or minutes are a direct consequences of time zone offsets; differences in year, month or day happen when you're close to the boundary and the offset makes you cross it.

This is why dates cannot be treated as a a less precise version of datetimes in general.

aaugustin replied:
In fact time zones affect Extract exactly like they affect Trunc (unless I misunderstood how this works).

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+    The ``week_day`` ``lookup_type`` is calculated differently from most
+    databases and from Python standard functions. This function will return
+    ``1`` for a Sunday, ``2`` for a Monday, through to ``7`` for a Saturday.
+
```
Perhaps you could explain why -- if you know the reason?

jarshwah replied:
The calculation was committed back in the early days (I looked it up, but can't find the commit just now) by Malcolm. There doesn't appear to be a very good reason other than "make all databases return the same value". I was thinking of adding an IsoWeekDay in the future, but it might be a good idea to add it now, and reference that from the docs here. Unfortunately, we'd be stuck with the original being registered as a week_day transform.

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+    1
+    >>> # Can be used on the right hand side of a filter
+    >>> Experiment.objects.filter(
+    ...    start_time__year=ExtractYear('start_time')).count()
```
This isn't a particularly interesting example... It always returns all results.

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+    .. attribute:: lookup_name = 'hour'
+
+Logically equivalent to ``Extract('date_field', 'hour')``. ``ExtractHour`` is
+also a ``Transform`` registered on Date and DateTime fields as ``__hour``.
```
ExtractHour/Minute/Day must not be registered on DateField.

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+
+.. class:: Trunc(date_expression, kind, output_field=None, **extra)
+
+Truncates a date up to a significant component.
```
"a date or a datetime" ?

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+When you only care if something happened in a particular year, hour, or day,
+but not the exact moment in time, then ``Trunc`` (and its subclasses) can be
+useful to filter or group your data. Calculating the number of sales per day
+is a common use case for ``Trunc``.
```
While one can reasonably describe truncation of datetimes to hours / minutes / seconds as loss of precision, further truncation of datetimes and all truncation of dates should really be described as aggregation.

That's pretty much what you did there ("group your data"). :+1: 

If I really wanted to be annoying, I'd object to using "moment in time" to describe (possibly) a date because a date is a calendaring concept only loosely related to a moment in time. At this precise moment in time, it's 2016-01-11 for me and 2016-04-12 for you ;-)

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+``DateTimeField``, a ``kind`` representing a date part, and an ``output_field``
+that's either ``DateTimeField()`` or ``DateField()``. It returns a datetime or
+date, depending on ``output_field``, with fields up to ``kind`` set to their
+minimum value.
```
``output_field`` should default to the type of the original field.

It isn't clear to me that truncating aware datetimes and getting dates as a result can be a valid operation. I think you must return always return aware datetimes in that case.

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+Logically equivalent to
+``Trunc('date_field', 'year', output_field=DateField())``. It truncates all
+parts of the date up to the year and allows grouping or filtering data within
+the same year. The return type is always ``DateField``.
```
That's bad when `USE_TZ = True` and one is truncating datetimes. In that case the result depends on the current time zone in effect at the time the query ran. Returning dates (instead of datetimes) looses that information.

aaugustin replied:
To put it differently: there's at least 20 intervals of time -- or sets of datetimes -- one could reasonably treat as "year 2016", depending on where one lives.

Outputting aware datetimes when the input is an aware datetimes clarifies which of these "years" was used.

aaugustin replied:
Obviously the same comment applies to month and date.

jarshwah replied:
This decision was consciously made because of the way timezones are applied *after* truncation. If I truncate `2016-06-01 00:00:00` to year, it'll return `2016-01-01 00:00:00`, apply a timezone offset of 1 hour, and the user receives `2015-12-31 23:00:00`. I'd find that result very surprising.

If you're already truncating to a date-part rather than a datetime-part, is there any value (or a use case) in retaining timezone information? Usually you're aggregating

jarshwah replied:
Whoa forget the above, I was totally wrong. You still get `2016-01-01 00:00:00`, just in an offset that's different to the one your started with to account for DST. I need to spend more time with these review notes before commenting on TZ behaviour.

aaugustin commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,518 @@ Usage example::
+    .. attribute:: output_field = DateTimeField()
+
+Logically equivalent to
+``Trunc('date_field', 'hour', output_field=DateTimeField())``. It truncates all
```
`'datetime_field'` ?

aaugustin commented on tests/db_functions/test_dates.py:
```diff
@@ -0,0 +1,611 @@
+        if kind == 'second':
+            return value.replace(microsecond=0)
+        if kind == 'minute':
+            return value.replace(second=0, microsecond=0)
```
I /think/ that's correct and doesn't require normalization because time zone offsets are integer number of minutes (barring pathological historical cases).

aaugustin commented on tests/db_functions/test_dates.py:
```diff
@@ -0,0 +1,611 @@
+        if kind == 'minute':
+            return value.replace(second=0, microsecond=0)
+        if kind == 'hour':
+            return value.replace(minute=0, second=0, microsecond=0)
```
I see two potential problems here:

1) If `value` is aware, this requires normalization in case the change makes you hit the DST threshold.

2) The correct semantic for the database is to operate in the current time zone. If this function is expected to do the same thing, the current implementation is insufficient, because some time zone offsets aren't integer number of hours.

I'm not sure what this function is used for, so perhaps this doesn't matter.

jarshwah replied:
This function is only used in the tests to make it easier to mimic what the database will return. I'll go back over it and check that there's no glaring inconsistencies. From memory I covered timezone conversions correctly.

(Melbourne is +10 now, but +11 at the start of the year)

trunc_year('2016-04-12 00:00:00', 'Melbourne') -> '2016-01-01 00:00:00' -> '2016-01-01 00:00:00 +11:00'

aaugustin commented on django/db/models/functions/date.py:
```diff
@@ -0,0 +1,214 @@
+class TruncBase(Transform):
+    arity = 1
+    kind = None
+    tzinfo = None
```
It's unclear to me why `TruncBase` has a `tzinfo` attribute while `Extract` doesn't have one.

I expect the two operations to have identical requirements and behavior when it comes to time zone handling.

aaugustin replied:
Wild guess -- `tzinfo` might be needed to store the current time zone at the time a call to QuerySet.datetimes() is made, as opposed to the time at which the resulting queryset is evaluated.

aaugustin commented on django/db/models/functions/date.py:
```diff
@@ -0,0 +1,214 @@
+                value = value.replace(tzinfo=None)
+                value = timezone.make_aware(value, self.tzinfo)
+        elif isinstance(self.output_field, DateField) and isinstance(value, datetime):
+            value = value.date()
```
These two lines make me sad... I had managed to delineate dates and datetimes cleanly by separating QuerySet.dates() and QuerySet.datetimes(). This new code makes it possible to accidentially treat datetimes and dates, which is usually signals incorrect database modeling and bugs waiting to happen (especially for Europeans who can ignore the different between <local time>.date() and <UTC>.date() as long as nothing happens within two hours of midnight, which is often the case for business appps).

aaugustin replied:
Actually I realize Django wasn't 100% clean in this regard. You merely moved some of the offending code. `datetime_cast_date_sql` isn't new.

aaugustin commented on tests/queries/tests.py:
```diff
@@ -1314,7 +1314,7 @@ def test_ticket7107(self):
         # An error should be raised when QuerySet.datetimes() is passed the
         # wrong type of field.
-        with self.assertRaisesMessage(AssertionError, "'name' isn't a DateTimeField."):
+        with self.assertRaisesMessage(AssertionError, "'name' isn't a DateField or DateTimeField."):
```
😢

jarshwah replied:
Other than this specific line, I believe all concerns have been addressed. This particular change was as a result of removing the specific DateTime and Date expressions, and switching .dates() and .datetimes() to using the generic `Trunc` class.

aaugustin replied:
If I understand correctly, this test shows that `queryset.datetimes()` now accepts a `DateField`.

We should check why I introduced this new method (in 1.6 if memory serves). Generally speaking the idea was that `dates()` should only accept dates and that `datetimes()` should only accept datetimes. I'm not sure why I couldn't do it just with the `dates()` API and that's what we need to check to determine the correct course of action.

jarshwah replied:
Good catch. I'll write a test to ensure this cannot happen. If the current API does allow dates into datetimes() then I can just query _meta for the field and do the check in the frontend method.

I haven't checked why you wrote two methods, but I suspect it's because tzinfo is only available on the datetimes() method. I'll double check tomorrow, as I've still got a couple of things to tidy up.

jarshwah commented:
Thanks for the review Tim and Aymeric. I was going to try and get most of this corrected in my first hour or two at $work$ but there's a lot here to go over. I'll get to all of this as soon as I can.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -364,6 +381,81 @@ Usage example::
+        'hour': 23, 'minute': 30, 'second': 1
+    }
+
+    >>> # When a user has a different timezone activated then results will
```
To break up the large code block, you could convert these comments into a regular paragraph.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -364,6 +381,81 @@ Usage example::
+    >>> Experiment.objects.create(
+    ...    start_time=start_2015, start_date=start_2015.date(),
+    ...    end_time=end_2015, end_date=end_2015.date())
+    >>> print(
```
Do you need to `print()` since the shell session automatically displays the result?

timgraham commented:
Needs a rebase and fixing undefined things so that the tests run. I'll review the docs tomorrow.

charettes commented:
Josh, do you think it could be worth mentioning `Date()` and `DateTime()` removal in the release note even if they were private API or actually implementation details for `dates()` and `datetimes()`?

jarshwah commented:
Probably a good idea Simon. I know some people look through the expressions file and use things there that aren't documented, so let's give those people a heads up.

jarshwah commented on django/db/models/functions/datetime.py:
```diff
@@ -0,0 +1,229 @@
+        # mistake.
+        if type(field) == DateField and (
+                isinstance(copy.output_field, DateTimeField) or copy.kind in ('hour', 'minute', 'second')):
+            raise ValueError("Attempting to convert DateTime %r to DateTimeField. " % field.name)
```
@aaugustin this is what I came up with to prevent dates being passed to functions expecting datetimes. The error is now a ValueError rather than an AssertionError, but there were no tests present to enforce this for `datetimes()`.

It seemed better to enforce this invariant in Trunc rather than the frontend .datetimes() queryset method. That, and using meta to traverse related fields to query types in the frontend method seemed like overkill 😈 

The obvious problem is that any datetime `kinds` that aren't explicitly supported by django do *not* enforce this restriction. Not a problem for `dates()` and `datetimes()` which do restrict the set of `kinds` available, and django can't know the actual expected type for unsupported kinds anyway.

I think this is a reasonable solution. I'm not entirely happy with the error message so I'm keen to hear advice on that aspect for anyone doing review.

timgraham replied:
Perhaps "Cannot convert DateField '%s' to DateTimeField."
(I'm not certain it's an issue here, but we're favoring '%s' over %r until we drop Python 2 support to avoid possible u'' prefixes.)

claudep commented on docs/releases/1.10.txt:
```diff
@@ -402,6 +402,17 @@ Models
 
 * A proxy model may now inherit multiple proxy models that share a common
   non-abstract parent class.
+* :class:`~django.db.models.functions.datetime.Extract` ``Func``-expression and
```
Newline missing above (often a side-effect of rebases).

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,391 @@ Usage example::
+
+The ``kind``\s above all have corresponding ``Trunc`` subclasses that should be
+preferred in the majority of cases. The subclasses below are all defined as
+transforms, but they are not registered with any fields by default.
```
say why not?

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -293,3 +293,391 @@ Usage example::
+``output_field`` of ``expression``. A ``tzinfo`` subclass, usually provided by
+``pytz``, can be passed to truncate a value in a specific timezone.
+
+Given the datetime `2015-06-15 14:30:50.000321+00:00` the ``kind``\s supported
```
This datetime has more significant digits in the fractional seconds part that the other "Given the datetime" above. If this isn't important or intentional, please make them consistent.

timgraham commented:
Many of the new tests fail on SQLite without pytz installed (`ImproperlyConfigured`) -- I didn't check on other databases. I guess the test skipping logic isn't quite right.

We also need to check on Windows without pytz `@requires_tz_support`, might be needed -- I can help if you don't have access to Windows (but the `ImproperlyConfigured` issue needs to be fixed first).

timgraham commented on tests/db_functions/test_datetime.py:
```diff
@@ -0,0 +1,684 @@
+            lambda m: (m.start_datetime, m.extracted)
+        )
+
+        result = 1 if connection.features.supports_microsecond_precision else 2
```
Something not quite right here as you have this condition but then only use `result` if `connection.features.supports_microsecond_precision`.

timgraham commented:
Please check my edits https://github.com/jarshwah/django/pull/3 as I haven't reviewed them carefully.
Also I noticed some failures on Oracle and edited the checklist in the PR description with details.
My review is completed.

jarshwah commented:
buildbot, test on oracle.

jarshwah commented:
Thanks for review Tim! There was a bug in oracles handling of second truncation when tz support was off. I've included the fix as a separate commit to highlight it incase of problems in the future.

I haven't got a windows machine to test on, but I think the decorators applied to the `DateFunctionWithTimeZoneTests` will sort out windows. The sqlite errors were sorted by skipping tests with the following decorator:

```
@skipIf(connection.vendor == 'sqlite' and pytz is None, "timezone support on sqlite requires pytz")
```

I couldn't find an existing decorator, feature, or test util that covered sqlite with timezone support enabled without pytz installed. Would it be worth creating a feature for this test case? I know we try to avoid vendor switching where possible.

The feature could be called something like `timezone_support_requires_pytz`? Thoughts anyone else following along?

jarshwah commented:
buildbot, test on oracle.

timgraham commented on tests/db_functions/test_datetime.py:
```diff
@@ -0,0 +1,642 @@
+            list(DTModel.objects.annotate(truncated=TruncSecond('start_date', output_field=DateField())))
+
+
+@requires_tz_support
```
`@requires_tz_support` skips the test on Windows. Actually these tests work on Windows as long as pytz in installed (tested with SQLite and PostgreSQL).

Without pytz installed on Windows:
SQLite: `ImproperlyConfigured: This query requires pytz, but it isn't installed.` (as on Linux)
PostgreSQL: `DataError: time zone "Eastern Standard Time" not recognized` (this situation works on Linux)
MySQL: `ValueError: Database returned an invalid datetime value. Are time zone definitions for your database and pytz installed?` for trunc tests and assertion errors like `(datetime.datetime(2015, 6, 15, 18, 30, 50, 321, tzinfo=<UTC>), None) != (datetime.datetime(2015, 6, 15, 14, 30, 50, 321, tzinfo=<django.utils.timezone.LocalTimezone object at 0x03C73A10>), 15)` for the extract tests. (again, no problem on Linux)

timgraham replied:
Rather than writing some more complex skip condition, I think I'm inclined to skip these tests if pytz isn't installed.

jarshwah replied:
I think you're right. It looks like Postgres + *nix is really the only platform that doesn't require pytz for basic timezone handling. For that reason it's probably just better to skip. I looked for documentation that compares databases + os's and their requirement for pytz but no such document exists.

I'll open a new ticket to address that separately.

I'll attach a requires pytz skip to this test class and call it a day. Thanks for your help with Windows!

timgraham commented on django/db/models/functions/datetime.py:
```diff
@@ -12,31 +12,60 @@
+        return tzname
+
+
+class Extract(Transform, TimezoneMixin):
```
I think mixins usually go on the left of the base class?

jarshwah replied:
Unless there's a style reason there's no need to put it to the left. The mixin isn't overriding anything so the MRO isn't an issue. I find the main class (Transform in this case) being on the left more readable.

claudep replied:
I second Tim. Unless a very specific use case, mixins should be on the left.

timgraham replied:
The code I've reviewed recently always has mixins on the left. I think it's easier to be consistent than have to think about it, but feel free to propose some different guidelines if you like.

jarshwah replied:
Yep fair enough, will make the change.

claudep replied:
Read also https://www.ianlewis.org/en/mixins-and-python

timgraham commented on django/db/models/functions/datetime.py:
```diff
@@ -12,31 +12,60 @@
             sql = connection.ops.time_extract_sql(self.lookup_name, sql)
         else:
-            raise ValueError('Extract only valid on Date/Time/DateTimeFields')
+            assert False, "Tried to Extract from an invalid type."
```
I'd rather raise an exception to account for the fact that assertions are ignored with `python -O` unless there's some reason to prefer the assert.

jarshwah replied:
This line should never be hit because resolve_expression below already validates the output_field. I'd prefer it to be an assertion for that reason.

timgraham replied:
Okay, please update the message or add a comment about that.

timgraham commented on django/db/models/functions/datetime.py:
```diff
@@ -12,31 +12,60 @@
+        copy = super(Extract, self).resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        field = copy.lhs.output_field
+        if not isinstance(field, (DateField, DateTimeField, TimeField)):
+            raise ValueError('Extract input expression must be one of DateField, DateTimeField or TimeField.')
```
chop "one of"
add comma before "or"

timgraham commented on django/db/models/functions/datetime.py:
```diff
@@ -12,31 +12,60 @@
+        # Passing dates to functions expecting datetimes is most likely a mistake.
+        if type(field) == DateField and copy.lookup_name in ('hour', 'minute', 'second'):
+            raise ValueError(
+                "Cannot extract time component '%s' from DateField '%s'. " % (copy.lookup_name, field.name))
```
put closing parenthesis on the next line

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -489,6 +491,21 @@ values that are returned::
     {'day': 16, 'weekday': 3, 'hour': 9}
 
+Explicitly passing the timezone to the ``Extract`` function behaves in the same
+way, and takes priority over an activated timezone::
```
activated -> active

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -489,6 +491,21 @@ values that are returned::
+    >>> Experiment.objects.annotate(
+    ...     day=ExtractDay('start_time', tzinfo=melb),
+    ...     weekday=ExtractWeekDay('start_time', tzinfo=melb),
+    ...     hour=ExtractHour('start_time', tzinfo=melb)
```
add trailing comma (missing in previous snippet too)

timgraham commented on tests/db_functions/test_datetime.py:
```diff
@@ -147,9 +147,14 @@ def test_extract_func(self):
 
-        with self.assertRaisesMessage(ValueError, 'Extract only valid on Date/Time/DateTimeFields'):
+        with self.assertRaisesMessage(
+                ValueError, 'Extract input expression must be one of DateField, DateTimeField or TimeField.'):
```
I usually write `msg = '...'`on the line before `with self.assertRaisesMessage(ValueError, msg):` to avoid the unusual indentation.

timgraham commented:
For the Oracle commit, perhaps you could add a sentence to the commit message like, "The tests for this change are added in the the fix for refs #25774." You could simplify the message on the first line a bit: " Made Oracle truncate microseconds if USE_TZ=False."

timgraham commented:
Oops, forgot to edit the exception messages in the tests.

There's a typo in the first commit message. Could you use this one:

Refs #25774 -- Made Oracle truncate microseconds if USE_TZ=False.

The tests for this change are in the fix for #25774.

jarshwah commented:
Merged in 2a4af0ea43512370764303d35bc5309f8abce666 and 77b73e79a4750dcbfabc528bf00cad81ff5bb4d9
